components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -120,6 +120,25 @@ class GeckoEngine(
         }
     }
 
+    /**
+     * Add support for service workers requests to open new tabs.
+     *
+     * @param onAddTab - callback to adding a tab using the provided [EngineSession].
+     * Implementations are expected to add a new normal (not private) tab but not load any URL in that.
+     * The background service worker will use the same passed in [EngineSession] to load it's needed URL.
+     */
+    fun installServiceWorkersSupport(
```

Comment:
Without adding this to the `Engine` API, in `feature-webnotifications` (or any other component) wouldn't you need to cast to a `GeckoEngine` to use it?

Reply:
Good point about this not being part of the `Engine` interface.
Can add it there also if you agree with the approach here - not a big fan of this implementation but couldn't find one better.

Reply:
What parts of the implementation do you not like here?

I think this is a fine direction - you could then make a `GeckoServiceWorkerDelegate` that implements an AC `ServiceWorkerDelegate` like the other delegate examples in this file.

Reply:
Tried extracting the behavior to a separate class delegate. Could not get it to work. The delegate has to be in this same package to have access to `GeckoEngineSession.geckoSession` which has `internal` visibility.

And then has to get passed to the `runtime` + `default settings`. Maybe from this call in `GeckoEngine`?  In Fenix `defaultSettings` are used in place. So I'd just complicate things with no real gain.

The part about `onAddTab` delegating this functionality to the client may seem nice but I was forced to - can't (and I think shouldn't) open tabs from the engine. But the client has no real control over that callback. With this implementation we are expecting it to really add a tab. Tried having that callback return a boolean but this `installServiceWorkersSupport` is set in Fenix - `Core.kt` which I didn't think is suited to have logic about whether to open tabs or not. 
I think I tried having a new feature just for this functionality but then I'd get errors about circular dependencies between the new module that would use `onAddTab` and the `browser-engine` module.